GitDiffMargin/Git/GitCommands.cs

```diff
@@ -282,5 +325,90 @@ static bool HasPreamble(string file, Encoding encoding)
+            switch (Path.GetExtension(fullPath).ToLowerInvariant())
+            {
+            case ".csproj":
+            case ".vbproj":
+                break;
+
+            default:
+                return fullPath;
+            }
+
+            // CPS will open the file in %TEMP%\{random name}\{ProjectFileName}
+            string directoryName = Path.GetDirectoryName(fullPath);
+            if (string.IsNullOrEmpty(directoryName))
+                return fullPath;
+
+            directoryName = Path.GetDirectoryName(directoryName);
+            if (!Path.GetTempPath().Equals(directoryName + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase))
+                return fullPath;
+
+            IVsSolution solution = _serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution;
+            if (solution == null)
+                return fullPath;
+
+            if (!ErrorHandler.Succeeded(solution.GetProjectEnum((uint)__VSENUMPROJFLAGS.EPF_LOADEDINSOLUTION, Guid.Empty, out IEnumHierarchies ppenum))
+                || ppenum == null)
+            {
+                return fullPath;
+            }
+
+            List<string> projectFiles = new List<string>();
+            IVsHierarchy[] hierarchies = new IVsHierarchy[1];
+            while (true)
+            {
+                int hr = ppenum.Next((uint)hierarchies.Length, hierarchies, out uint fetched);
+                if (!ErrorHandler.Succeeded(hr))
+                    return fullPath;
+
+                for (uint i = 0; i < fetched; i++)
+                {
+                    if (!(hierarchies[0] is IVsProject project))
+                        continue;
+
+                    if (!ErrorHandler.Succeeded(project.GetMkDocument((uint)VSConstants.VSITEMID.Root, out string projectFilePath)))
+                        continue;
+
+                    if (!Path.GetFileName(projectFilePath).Equals(Path.GetFileName(fullPath), StringComparison.OrdinalIgnoreCase))
+                        continue;
+
+                    projectFiles.Add(projectFilePath);
+                }
+
+                if (hr != VSConstants.S_OK)
+                {
+                    // No more projects
+                    break;
+                }
+            }
+
+            switch (projectFiles.Count)
+            {
+            case 0:
+                // No matching project file found in solution
+                return fullPath;
+
+            case 1:
+                // Exactly one matching project file found in solution
+                return projectFiles[0];
+
+            default:
+                // Multiple project files found in solution; try to find one with a matching file size
+                long desiredSize = new FileInfo(fullPath).Length;
+                foreach (var projectFilePath in projectFiles)
+                {
+                    if (File.Exists(projectFilePath) && new FileInfo(projectFilePath).Length == desiredSize)

```

Comment:
> can you please test the following case

The detection of the original file path worked fine here, but the `IsDirty` check failed, resulting in an empty diff. I corrected this as its own commit in the branch. 👍